### PR TITLE
Cluster heat nodes

### PR DESCRIPTION
**Pull request by lisazeyen**
This PR should add the functionality to optional cluster residential and service heat buses. This saves especially for larger networks memory. Tackles issue #215

### CODE REVIEW

File: scripts/prepare_sector_network.py [hunk 1]
```diff
@@ -19,13 +19,17 @@
...
 from pypsa.geo import haversine_pts
+from pypsa.io import import_components_from_dataframe
 
 import logging
 logger = logging.getLogger(__name__)
 
 from types import SimpleNamespace
 spatial = SimpleNamespace()
 
+from distutils.version import LooseVersion
```
Comment (fneum) [inline scripts/prepare_sector_network.py:30]: This is now deprecated. Use packaging tools instead:

https://github.com/PyPSA/PyPSA/blob/6bf79d8f7a7e01c6cac6668dc0675f33fa90b29b/pypsa/networkclustering.py#L23


https://github.com/PyPSA/PyPSA/blob/6bf79d8f7a7e01c6cac6668dc0675f33fa90b29b/pypsa/networkclustering.py#L744

Comment (lisazeyen) [reply scripts/prepare_sector_network.py]: thanks, I changed it now!

**Comment by lisazeyen** [comment]
@energyLS could you have a look?

**Review by fneum — APPROVED** [review]

File: scripts/add_existing_baseyear.py [hunk 2]
```diff
@@ -495,4 +495,7 @@ def add_heating_capacities_installed_before_baseyear(n, baseyear, grouping_years
         default_lifetime = snakemake.config['costs']['lifetime']
         add_heating_capacities_installed_before_baseyear(n, baseyear, grouping_years, ashp_cop, gshp_cop, time_dep_hp_cop, costs, default_lifetime)
 
+    if options["cluster_heat_buses"]:
```
Comment (fneum) [inline scripts/add_existing_baseyear.py:498]:
```suggestion
    if options.get("cluster_heat_buses", False):
```

File: scripts/prepare_sector_network.py [hunk 3]
```diff
@@ -2467,4 +2564,10 @@ def limit_individual_line_extension(n, maxext):
     if options['electricity_grid_connection']:
         add_electricity_grid_connection(n, costs)
 
+    first_year_myopic = ((snakemake.config["foresight"] == 'myopic') and
+                         (snakemake.config["scenario"]["planning_horizons"][0]==investment_year))
+
+    if options["cluster_heat_buses"] and not first_year_myopic:
```
Comment (fneum) [inline scripts/prepare_sector_network.py:2570]:
```suggestion
    if options.get("cluster_heat_buses", False) and not first_year_myopic:
```

File: config.default.yaml [hunk 4]
```diff
@@ -160,6 +160,7 @@ sector:
       2040: 0.6
       2050: 1.0
     district_heating_loss: 0.15
+  cluster_heat_buses: False # cluster residential and service heat buses to one to save memory
```
Comment (fneum) [inline config.default.yaml:163]:
```suggestion
  cluster_heat_buses: false # cluster residential and service heat buses to one to save memory
```

**Review by fneum — APPROVED** [review]

**Comment by fneum** [comment]
Great @lisazeyen! Can be merged!

**Pull request merged**